### PR TITLE
Optim tests run for multiple steps

### PR DESCRIPTION
**Pull request by TomFinet**

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -52,23 +42,29 @@ def forward(self):
     return out
 
 
+def helper_test_optim(steps, tinyoptim, torchoptim, tinyargs={}, torchargs={}):
+  tinynet, torchnet = TinyNet(), TorchNet()
+  for _ in range(steps):
+    step(tinynet, tinyoptim, tinyargs)
+    step(torchnet, torchoptim, torchargs)
+    x1, y1 = tinynet.x.cpu().data, tinynet.W.cpu().data
```
Comment (geohot) [inline test/test_optim.py:50]: .detach().numpy() should work for both now

File: test/test_optim.py [hunk 2]
```diff
@@ -52,23 +42,29 @@ def forward(self):
...
-                   step_pytorch(torch.optim.Adam)):
-      np.testing.assert_allclose(x, y, atol=1e-4)
+    helper_test_optim(steps, Adam, torch.optim.Adam)
 
   def test_sgd(self):
-    for x,y in zip(step_tinygrad(SGD, kwargs={'lr': 0.001}),
-                   step_pytorch(torch.optim.SGD, kwargs={'lr': 0.001})):
-      np.testing.assert_allclose(x, y, atol=1e-5)
+    helper_test_optim(steps, SGD, torch.optim.SGD, 
+                      tinyargs={'lr': 0.001}, torchargs={'lr': 0.001})
```
Comment (geohot) [inline test/test_optim.py:63]: Add SGD with momentum/nesterov tests?

File: test/test_optim.py [hunk 3]
```diff
@@ -52,23 +42,29 @@ def forward(self):
...
+  tinynet, torchnet = TinyNet(), TorchNet()
+  for _ in range(steps):
+    step(tinynet, tinyoptim, tinyargs)
+    step(torchnet, torchoptim, torchargs)
+    x1, y1 = tinynet.x.cpu().data, tinynet.W.cpu().data
+    x2, y2 = torchnet.x.detach().numpy(), torchnet.W.detach().numpy()
+    np.testing.assert_allclose(x1, x2, atol=1e-4)
+    np.testing.assert_allclose(y1, y2, atol=1e-4)
+
+steps = 3
```
Comment (geohot) [inline test/test_optim.py:55]: Still work with 10?

**Comment by geohot** [comment]
Hmm, tests failing

**Comment by TomFinet** [comment]
Only plain SGD fails.

**Comment by TomFinet** [comment]
If we use the following SGD optim class, all tests pass for any number of steps.

`class SGD(Optimizer):
  def __init__(self, params : List[Tensor], lr=0.001, momentum=0, nesterov=False):
    super().__init__(params)
    self.lr, self.momentum, self.nesterov = lr, momentum, nesterov
    self.b = [Tensor.zeros(*t.shape, device=params[0].device, requires_grad=False) for t in self.params] #if self.momentum else []

  # https://pytorch.org/docs/stable/generated/torch.optim.SGD.html
  def step(self) -> None:
    for i, t in enumerate(self.params):
      assert t.grad is not None
      g = t.grad
      #if self.momentum:
      self.b[i].assign(self.momentum * self.b[i] + g)
      g = (g + self.momentum * self.b[i]) if self.nesterov else self.b[i]
      t.assign(t.detach() - g * self.lr)
    self.realize(self.b)`

**Comment by TomFinet** [comment]
If we remove `if self.momentum` from the SGD class step method, and always run the momentum code with momentum=0.0 when we don't want it, then all tests pass for any number of steps (on my machine).

**Comment by TomFinet** [comment]
nevermind

**Comment by geohot** [comment]
Please reopen if you fix it
